module/zfs/dmu.c

```diff
@@ -1387,628 +2307,290 @@ dmu_read_uio_dbuf(dmu_buf_t *zdb, uio_t *uio, uint64_t size)
-		dr->dt.dl.dr_copies = zio->io_prop.zp_copies;
+	for (int i = 0; i < numbufs; i++) {
+		uint64_t tocpy;
+		int64_t bufoff;
+		dmu_buf_t *db = dbp[i];
 
-		/*
-		 * Old style holes are filled with all zeros, whereas
-		 * new-style holes maintain their lsize, type, level,
-		 * and birth time (see zio_write_compress). While we
-		 * need to reset the BP_SET_LSIZE() call that happened
-		 * in dmu_sync_ready for old style holes, we do *not*
-		 * want to wipe out the information contained in new
-		 * style holes. Thus, only zero out the block pointer if
-		 * it's an old style hole.
-		 */
-		if (BP_IS_HOLE(&dr->dt.dl.dr_overridden_by) &&
-		    dr->dt.dl.dr_overridden_by.blk_birth == 0)
-			BP_ZERO(&dr->dt.dl.dr_overridden_by);
-	} else {
-		dr->dt.dl.dr_override_state = DR_NOT_OVERRIDDEN;
-	}
-	cv_broadcast(&db->db_changed);
-	mutex_exit(&db->db_mtx);
+		ASSERT(size > 0);
 
-	dsa->dsa_done(dsa->dsa_zgd, zio->io_error);
+		bufoff = uio->uio_loffset - db->db_offset;
+		tocpy = MIN(db->db_size - bufoff, size);
 
-	kmem_free(dsa, sizeof (*dsa));
-}
+		ASSERT(i == 0 || i == numbufs-1 || tocpy == db->db_size);
 
-static void
-dmu_sync_late_arrival_done(zio_t *zio)
-{
-	blkptr_t *bp = zio->io_bp;
-	dmu_sync_arg_t *dsa = zio->io_private;
-	zgd_t *zgd = dsa->dsa_zgd;
+		if (tocpy == db->db_size)
+			dmu_buf_will_fill(db, tx);
+		else
+			dmu_buf_will_dirty(db, tx);
 
-	if (zio->io_error == 0) {
 		/*
-		 * Record the vdev(s) backing this blkptr so they can be
-		 * flushed after the writes for the lwb have completed.
+		 * XXX uiomove could block forever (eg.nfs-backed
+		 * pages).  There needs to be a uiolockdown() function
+		 * to lock the pages in memory, so that uiomove won't
+		 * block.
 		 */
-		zil_lwb_add_block(zgd->zgd_lwb, zgd->zgd_bp);
+		err = uiomove((char *)db->db_data + bufoff, tocpy,
+		    UIO_WRITE, uio);
 
-		if (!BP_IS_HOLE(bp)) {
-			blkptr_t *bp_orig __maybe_unused = &zio->io_bp_orig;
-			ASSERT(!(zio->io_flags & ZIO_FLAG_NOPWRITE));
-			ASSERT(BP_IS_HOLE(bp_orig) || !BP_EQUAL(bp, bp_orig));
-			ASSERT(zio->io_bp->blk_birth == zio->io_txg);
-			ASSERT(zio->io_txg > spa_syncing_txg(zio->io_spa));
-			zio_free(zio->io_spa, zio->io_txg, zio->io_bp);
-		}
-	}
+		if (tocpy == db->db_size)
+			dmu_buf_fill_done(db, tx);
 
-	dmu_tx_commit(dsa->dsa_tx);
+		if (err)
+			break;
 
-	dsa->dsa_done(dsa->dsa_zgd, zio->io_error);
+		size -= tocpy;
+	}
 
-	abd_put(zio->io_abd);
-	kmem_free(dsa, sizeof (*dsa));
+	dmu_buf_rele_array(dbp, numbufs, FTAG);
+	return (err);
 }
 
-static int
-dmu_sync_late_arrival(zio_t *pio, objset_t *os, dmu_sync_cb_t *done, zgd_t *zgd,
-    zio_prop_t *zp, zbookmark_phys_t *zb)
+/*
+ * Write 'size' bytes from the uio buffer.
+ * To object zdb->db_object.
+ * Starting at offset uio->uio_loffset.
+ *
+ * If the caller already has a dbuf in the target object
+ * (e.g. its bonus buffer), this routine is faster than dmu_write_uio(),
+ * because we don't have to find the dnode_t for the object.
+ */
+int
```

Comment:
I'm not sure it's possible, but it would be helpful for reviewers if you could rearrange this code so that the diffs to existing functions show up as such, rather than showing all the old code deleted and then "new code" added (which is presumably almost the same as the old code).  If that isn't possible, perhaps you could do a first commit that only moves code around, and then a 2nd commit that adds directio support?

Reply:
Yes, I agree that it rather confusing, frustrating to try and sort out what is new code and what is just fragments due to rearranging the code. I will look into the easy way to accomplish this and update the PR accordingly.

Reply:
I went ahead and did a separate commit for the rearranging dmu.c, so now the WIP commit just shows additions to this file now. Definitely should make it easier to evaluate the changes only.